src/pytest_inmanta_lsm/lsm_project.py

```diff
@@ -926,81 +936,99 @@ def add_service(
+        # Get the service's current version
+        service = self.get_service(service_id)
 
-        if validation:
-            if len(service_ids) != 1:
-                raise Exception(
-                    f"when performing a validation compile, only one service id can be passed, got {repr(service_ids)}"
-                )
+        # Collect the environment variables that will need to be set for the compile
+        env = {
+            inmanta_lsm.const.ENV_INSTANCE_VERSION: str(service.version),
+            inmanta_lsm.const.ENV_MODEL_STATE: inmanta_lsm.model.ModelState.candidate.value,
+            inmanta_lsm.const.ENV_INSTANCE_ID: str(service_id),
+            inmanta_lsm.const.ENV_PARTIAL_COMPILE: str(self.partial_compile),
+        }
 
-            # Get the service's current version to set it in the env var
-            service = self.get_service(service_ids[0])
-            env[inmanta_lsm.const.ENV_INSTANCE_VERSION] = str(service.version)
+        LOGGER.debug(
+            "Triggering validating compile for service %s with the following environment variables: %s",
+            service_id,
+            env,
+        )
+        with pytest.MonkeyPatch.context() as m:
+            for k, v in env.items():
+                m.setenv(k, v)
+            self.project.compile(model, no_dedent=False)
 
-            # If we have a validation compile, we need to set an additional env var
-            env[inmanta_lsm.const.ENV_MODEL_STATE] = inmanta_lsm.model.ModelState.candidate
+    def exporting_compile(
+        self,
+        service_ids: collections.abc.Sequence[str | uuid.UUID] | None = None,
+        *,
+        model: str | None = None,
+    ) -> None:
+        """
+        Perform an exporting compile for the given services, or all services present in the inventory
+        if no services is specified.
+        Use the provided model for the compilation.  If none is provided, defaults to the model used
+        to export the service definition if there is any, fails otherwise.
+        """
+        # Get the model to use for the compile
+        model = self._get_model(model)
 
-        try:
-            env[inmanta_lsm.const.ENV_PARTIAL_COMPILE] = str(self.partial_compile and bool(service_ids))
-        except AttributeError:
-            # This attribute only exists for iso5+, iso4 doesn't support partial compile.
-            # We then simply don't set the value.
-            if self.partial_compile:
-                warnings.warn("Partial compile is enabled but it is not supported, it will be ignored.")
+        if service_ids is None:
+            # Triggering a full compile
+            LOGGER.debug("Triggering full compile")
+            self.project.compile(model, no_dedent=False)
+            return
+
+        if not self.partial_compile:
+            # Specific services are requested for the compile but partial compile is not
+            # supported, this is a user error
```

Comment:
What does not supported mean in this context?

i.e. if someone want to run a test in partial compile mode, what causes us to prevent that?

Reply:
Not supported whean that `LsmProject.partial_compile` is set to `False`, so a test can be run in partial compile mode, it just needs to be set on the class

Reply:
how is it set?

Reply:
The default value will depend on `INMANTA_LSM_PARTIAL_COMPILE` env var: https://github.com/inmanta/pytest-inmanta-lsm/blob/cb32891661e4cb93990ea444afe66d1eacb91bb7/src/pytest_inmanta_lsm/plugin.py#L110

But in a partial compile specific test, you will usually set it manually anyway: https://code.inmanta.com/solutions/modules/connect/-/blob/7b71db1e242243a12d2b9e151e1982198fbff752/tests/test_model/test_l2_connect.py#L73

Reply:
I find this confusing, but I'm not the primary user. 
I would at least mention in the exception how to resolve the problem.

Reply:
I updated it, is it more clear?